src/libraries/System.Private.CoreLib/src/System/ParamsArray.cs

```diff
@@ -7,8 +7,48 @@
 // Suppress warnings for unused private fields
 #pragma warning disable CS0169, CA1823, IDE0051, IDE0044
 
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
 namespace System
 {
+#if CORECLR || NATIVEAOT
```

Comment:
https://github.com/dotnet/runtime/blob/6d325e1aa46879677a56b722fc20ef919ec173f5/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBase.cs#L301 is another good candidate to use this if you are looking for real-world coverage.

Reply:
Also:

- https://github.com/dotnet/runtime/blob/3e05fa1e1b4d127bd9a2c9b2e7d1d8bcb2d7a2e8/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexReplacement.cs#L97-L104
- https://github.com/dotnet/runtime/blob/6d325e1aa46879677a56b722fc20ef919ec173f5/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventProvider.cs#L1030-L1042
- https://github.com/dotnet/runtime/blob/3e05fa1e1b4d127bd9a2c9b2e7d1d8bcb2d7a2e8/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObjectSequence.netcore.cs#L16-L37
- https://github.com/dotnet/runtime/blob/3e05fa1e1b4d127bd9a2c9b2e7d1d8bcb2d7a2e8/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/StringSequence.netcore.cs#L17-L38

Reply:
Thanks! the `ref struct ` one is particularly interesting. We take a different path in more than one place internally for `IsByRefLike` things. 
There is nothing wrong with value array containing refs or reflikes though, so there is some interaction of features and new code in those areas.

Reply:
Note that I am trying to do only simple changes here. The goal is just to get coverage for tricky layout situations. 
I know that this can be fancier. For example we may have an overloaded factory for params and do
```cs
foo(Params(a,b,c).AsSpan());
```
Or maybe Roslyn will do this automatically for `params Span<T> p`

It would be easier to introduce more conveniences as a separate change and it is probably worth waiting for Mono support, so that there is not too many `#if`s
 
For now I am trying to use the feature in CorLib, but also trying to limit the changes.

Reply:
I will have tests, but CorLib is interesting as it also goes through crossgen and that has a separate layout engine which should produce the same results as native type system.

Reply:
I also converted `ObjectSequence2` and friends to be an InlineArray. 
It worked, but I did not put it in since Diagnostics also builds for 6.0 and decided to not special case for that.